fix: Add PBAC permission checks for insights access

dhairyashiil opened the pull request:
Before:

https://github.com/user-attachments/assets/bb0cb00a-991f-4afb-843c-30f66094b060



After: 

https://github.com/user-attachments/assets/5c487a0e-b42f-4cf8-9cc4-826b450be119




## Problem

- Users in PBAC-enabled organizations couldn't access insights page
- Even admins with `admin_role` (which has `insights.read` permission) were blocked
- The code was only checking the base `role` field (ADMIN/OWNER/MEMBER)
- It completely ignored PBAC permissions

## What This Fixes

- Added proper PBAC permission checks using `PermissionCheckService`
- Admins with PBAC roles can now access insights
- Custom roles with `insights.read` permission now work
- Non-PBAC organizations still work the same way (backward compatible)

## Changes Made

1. Created `checkInsightsPermission()` helper function
   - Checks if user has `insights.read` permission
   - Uses PBAC system when enabled
   - Falls back to checking ADMIN/OWNER role when PBAC disabled

2. Updated organization-level access check
   - Now uses PBAC permission check instead of direct role query

3. Updated team list query
   - Filters teams based on `insights.read` permission
   - Works with both PBAC roles and traditional roles

## Testing

✅ PBAC enabled + `admin_role` → Can access insights 
✅ PBAC enabled + custom role with `insights.read` → Can access insights
✅ PBAC enabled + custom role without `insights.read` → Cannot access insights
✅ PBAC disabled + traditional ADMIN → Can access insights
✅ PBAC disabled + traditional MEMBER → Cannot access insights

dhairyashiil committed:
fix: Add PBAC permission checks for insights access

- Add checkInsightsPermission() helper that properly checks insights.read permission with PBAC support
- Update userBelongsToTeamProcedure to use PBAC-aware permission check for org-level access
- Update teamListForUser query to filter teams based on insights.read permission instead of only checking base ADMIN/OWNER roles
- Maintain backward compatibility with fallback to traditional role checks (ADMIN/OWNER) when PBAC is disabled
- Org admins (base role ADMIN/OWNER) continue to have automatic insights access as a privileged position
- Team-level admins with custom roles now properly checked for insights.read permission

Fixes issue where users with custom PBAC roles couldn't access insights even if they had insights.read permission.

Related: CAL-XXXX

Udit-takkar commented on packages/features/insights/server/trpc-router.ts:
```diff
@@ -685,9 +681,18 @@ export const insightsRouter = router({
+    );
+
+    const result: IResultTeamList[] = teamsWithAccess
+      .filter((team): team is NonNullable<typeof team> => team !== null)
```
I think we can use `getTeamIdsWithPermission` method in permission-service to avoid N+1 queries here

dhairyashiil committed:
perf: Optimize team permission checks to avoid N+1 queries

Replace individual permission checks per team with bulk query using getTeamIdsWithPermission().
This reduces database queries from N (one per team) to a single optimized query.

- Use PermissionCheckService.getTeamIdsWithPermission() for bulk permission checking
- Filter teams based on returned team IDs instead of individual checks
- Maintains same functionality with significantly better performance for users with many teams

sean-brydon commented on packages/features/insights/server/trpc-router.ts:
```diff
@@ -685,10 +681,20 @@ export const insightsRouter = router({
+    // Get all team IDs where user has insights.read permission in a single optimized query
+    // This avoids N+1 queries by fetching all permissions at once
+    const permissionCheckService = new PermissionCheckService();
+    const teamIdsWithAccess = await permissionCheckService.getTeamIdsWithPermission({
```
I don't believe so - will double check the code

sean-brydon replied:
Actually believe this to be the case - working on a fix. SQL will be pretty complex here so wll push a new PR

dhairyashiil committed:
perf: Fetch only teams with insights access instead of filtering after

Move permission check before team query to filter at database level.
Previously fetched all teams then filtered in JavaScript.
Now only fetches teams the user has insights access to.

- Check permissions first using getTeamIdsWithPermission()
- Add teamId filter to membership query (teamId: { in: teamIdsWithAccess })
- Remove JavaScript filter step (done at DB level)
- Reduces data transfer and improves query efficiency

Udit-takkar reviewed (approved):
LGTM

dhairyashiil committed:
Merge branch 'main' into fix/pbac-insights-permission-check
